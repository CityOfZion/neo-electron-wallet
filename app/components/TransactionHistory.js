import React, { Component } from 'react';
import { connect } from 'react-redux';
import { syncTransactionHistory } from "../components/NetworkSwitch";
import { shell } from 'electron';
import Copy from 'react-icons/lib/md/content-copy';
import { clipboard } from 'electron';
import ExplorerSwitch from '../components/ExplorerSwitch'
// TODO: make this a user setting
const getExplorerLink = (net, txid, explorer) => {
  let base;
  if (net === "MainNet"){
    base = explorer.mainnet_url;
  } else {
    base = explorer.testnet_url;
  }
  return `${base}${explorer.hash_path}${txid}`;
}

// helper to open an external web link
const openExplorer = (srcLink) => {
  shell.openExternal(srcLink);
}

class TransactionHistory extends Component {

  componentDidMount = () => {
    syncTransactionHistory(this.props.dispatch, this.props.net, this.props.address);
  }

  render = () =>
    <div id="transactionInfo">
      <div className="columnHeader">Transaction History</div>
      <ExplorerSwitch />
      <div className="headerSpacer"></div>
      <ul id="transactionList">
        {this.props.transactions.map((t) => {
          let formatAmount;
          if (t.type === "NEO"){ formatAmount = parseInt(t.amount); }
          else{ formatAmount = parseFloat(t.amount).toPrecision(5); }
          // ignore precision rounding errors for GAS
          if ((formatAmount > 0 && formatAmount < 0.001) || (formatAmount < 0 && formatAmount > -0.001)){
            formatAmount = 0.0.toPrecision(5);
          }
          return (<li key={t.txid}>
              <div className="txid" onClick={() => openExplorer(getExplorerLink(this.props.net, t.txid, this.props.explorer))}>
                {t.txid.substring(0,32)}</div><div className="amount">{formatAmount} {t.type}
              </div></li>);
        })}
      </ul>
    </div>;
}

const mapStateToProps = (state) => ({
  address: state.account.address,
  net: state.metadata.network,
  transactions: state.wallet.transactions,
  explorer: state.explorer.selected
});

TransactionHistory = connect(mapStateToProps)(TransactionHistory);

export default TransactionHistory;
