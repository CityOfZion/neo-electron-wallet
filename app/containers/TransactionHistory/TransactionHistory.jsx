import React from 'react'

import TransactionHistoryPanel from '../../components/TransactionHistory/TransactionHistoryPanel'

import styles from './TransactionHistory.scss'

export default function TransactionHistory() {
  return (
    <div className={styles.transactionHistory}>
      <TransactionHistoryPanel className={styles.transactionHistoryPanel} />
    </div>
  )
}
