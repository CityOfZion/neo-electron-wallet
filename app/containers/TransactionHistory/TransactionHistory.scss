.transactionHistory {
  display: flex;
  flex-direction: column;
  height: 100%;

  .transactionHistoryPanel {
    flex: 1 1 auto;
  }
}

.panelHeaderButtons {
  display: flex;
  justify-content: space-between;
}

.exportButton {
  font-size: 16px !important;

  svg {
    path {
      fill: var(--header-bar-default-icon-color) !important;
    }
  }
}
