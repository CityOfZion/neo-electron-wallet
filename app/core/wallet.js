// @flow
import { ASSETS, TOKENS } from './constants'
import { getAccountFromWIFKey, verifyAddress } from 'neon-js'

const MIN_PASSPHRASE_LEN = 4

export const validatePassphrase = (passphrase: string): boolean => passphrase.length >= MIN_PASSPHRASE_LEN

export const checkMatchingPassphrases = (passphrase: string, passphrase2: string) => passphrase !== passphrase2

export const verifyPrivateKey = (wif: string): boolean => {
  if (!wif) {
    return false
  }
  const account = getAccountFromWIFKey(wif)
  return account !== -1 && account.address
}

export const isToken = (symbol: TokenSymbol) => Object.keys(TOKENS).includes(symbol)

export const obtainTokenBalance = (tokens: Array<TokenType>, symbol: TokenSymbol) => {
  if (!isToken(symbol)) {
    throw new Error(`${symbol} is not a valid token`)
  }
  return tokens.find((token: TokenType) => token.symbol === symbol)
}

export const validateTransactionBeforeSending = (neoBalance: number, gasBalance: number, tokenBalance: number, symbol: TokenSymbol, sendAddress: string, sendAmount: string) => {
  if (!sendAddress || !sendAmount) {
    return {
      error: 'Please specify an address and amount',
      valid: false
    }
  }

  if (symbol !== ASSETS.NEO && symbol !== ASSETS.GAS && !isToken(symbol)) {
    return {
      error: 'That asset is not NEO, GAS or NEP-5 Token',
      valid: false
    }
  }

  try {
    if (verifyAddress(sendAddress) !== true || sendAddress.charAt(0) !== 'A') {
      return {
        error: 'The address you entered was not valid.',
        valid: false
      }
    }
  } catch (e) {
    return {
      error: 'The address you entered was not valid.',
      valid: false
    }
  }

  if (symbol === ASSETS.NEO) {
    if (parseFloat(sendAmount) !== parseInt(sendAmount)) { // check for fractional neo
      return {
        error: 'You cannot send fractional amounts of NEO.',
        valid: false
      }
    }
    if (parseInt(sendAmount) > neoBalance) { // check for value greater than account balance
      return {
        error: 'You do not have enough NEO to send.',
        valid: false
      }
    }
  } else if (symbol === ASSETS.GAS) {
    if (parseFloat(sendAmount) > gasBalance) {
      return {
        error: 'You do not have enough GAS to send.',
        valid: false
      }
    }
  } else {
    if (parseFloat(sendAmount) > tokenBalance) {
      return {
        error: `You do not have enough ${symbol} to send.`,
        valid: false
      }
    }
  }

  if (parseFloat(sendAmount) < 0) { // check for negative asset
    return {
      error: 'You cannot send negative amounts of an asset.',
      valid: false
    }
  }

  return {
    error: '',
    valid: true
  }
}
