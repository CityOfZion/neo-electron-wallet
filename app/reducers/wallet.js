import * as types from '../actions/types';

// reducer for wallet account balance
export default (state = {Neo: 0, Gas: 0, transactions: [], price: '--'}, action) => {
  switch (action.type) {
      case types.SET_BALANCE:
          return {...state, 'Neo': action.Neo, 'Gas': action.Gas, 'price': action.price };
      case types.RESET_PRICE:
          return {...state, 'price': '--'};
      case types.SET_MARKET_PRICE:  //current market price action type
          let currentPrice;
          if (action.price !== undefined){
              currentPrice = action.price;
          } else {
              currentPrice = '--';
          }
          return {...state, price: currentPrice};
      case types.SET_TRANSACTION_HISTORY:
          return {...state, transactions: action.transactions};
      default:
          return state;
  }
};