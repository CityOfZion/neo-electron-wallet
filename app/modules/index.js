// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import transactions from './transactions'
import wallet from './wallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  wallet,
  transactions,
  claim,
  notifications,
  modal,
  addressBook
})
