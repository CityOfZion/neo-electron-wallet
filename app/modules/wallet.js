// @flow
import { ASSETS, TOKENS } from '../core/constants'
import { merge } from 'lodash'
import axios from 'axios'
import { getBalance, getTokenBalance, getTokenInfo } from 'neon-js'
import { syncTransactionHistory } from './transactions'
import { syncAvailableClaim } from './claim'
import { syncBlockHeight, getNetwork } from './metadata'
import asyncWrap from '../core/asyncHelper'
import { LOGOUT, getAddress } from './account'

const TOKEN_PAIRS = Object.entries(TOKENS)

// Constants
export const SET_BALANCE = 'SET_BALANCE'
export const SET_NEO_PRICE = 'SET_NEO_PRICE'
export const SET_GAS_PRICE = 'SET_GAS_PRICE'
export const RESET_PRICE = 'RESET_PRICE'
export const SET_TRANSACTION_HISTORY = 'SET_TRANSACTION_HISTORY'
export const SET_TOKENS_BALANCE = 'SET_TOKENS_BALANCE'
export const SET_TOKENS_INFO = 'SET_TOKENS_INFO'

// Actions
export function setBalance (NEO: number, GAS: number) {
  return {
    type: SET_BALANCE,
    payload: { NEO, GAS }
  }
}

export function setNeoPrice (neoPrice: number) {
  return {
    type: SET_NEO_PRICE,
    payload: { neoPrice }
  }
}

export function setGasPrice (gasPrice: number) {
  return {
    type: SET_GAS_PRICE,
    payload: { gasPrice }
  }
}

export function resetPrice () {
  return {
    type: RESET_PRICE
  }
}

export function setTransactionHistory (transactions: Array<Object>) {
  return {
    type: SET_TRANSACTION_HISTORY,
    payload: { transactions }
  }
}

export function setTokensBalance (tokens: Array<Object>) {
  return {
    type: SET_TOKENS_BALANCE,
    payload: { tokens }
  }
}

export function setTokensInfo (tokensInfo: Array<Object>) {
  return {
    type: SET_TOKENS_INFO,
    payload: { tokensInfo }
  }
}

export const getMarketPriceUSD = () => async (dispatch: DispatchType) => {
  // If API dies, still display balance - ignore _err
  const [_err, response] = await asyncWrap(axios.get('https://api.coinmarketcap.com/v1/ticker/neo/?convert=USD')) // eslint-disable-line
  let lastUSDNEO = Number(response.data[0].price_usd)
  return dispatch(setNeoPrice(lastUSDNEO))
}

export const getGasMarketPriceUSD = () => async (dispatch: DispatchType) => {
  // If API dies, still display balance - ignore _err
  const [_err, response] = await asyncWrap(axios.get('https://api.coinmarketcap.com/v1/ticker/gas/?convert=USD')) // eslint-disable-line
  let lastUSDGAS = Number(response.data[0].price_usd)
  return dispatch(setGasPrice(lastUSDGAS))
}

export const retrieveBalance = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  // If API dies, still display balance - ignore _err
  const [_err, resultBalance] = await asyncWrap(getBalance(net, address)) // eslint-disable-line
  return dispatch(setBalance(resultBalance.NEO.balance, resultBalance.GAS.balance))
}

export const initiateGetBalance = (net: NetworkType, address: string) => (dispatch: DispatchType) => {
  dispatch(syncTransactionHistory(net, address))
  dispatch(syncAvailableClaim(net, address))
  dispatch(syncBlockHeight(net))
  dispatch(getMarketPriceUSD())
  dispatch(getGasMarketPriceUSD())
  dispatch(retrieveTokensBalance())
  dispatch(retrieveTokensInfo())
  return dispatch(retrieveBalance(net, address))
}

export const retrieveTokensBalance = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const net = getNetwork(state)
  const address = getAddress(state)
  const tokensBalance = []
  for (const [symbol, scriptHash] of Object.entries(TOKENS)) {
    let [_err, results] = await asyncWrap(getTokenBalance(net, scriptHash, address)) // eslint-disable-line
    if (results) {
      tokensBalance.push({
        symbol,
        balance: results
      })
    }
  }

  return dispatch(setTokensBalance(tokensBalance))
}

export const retrieveTokensInfo = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const net = getNetwork(state)
  const tokensInfo = []
  for (const [symbol, scriptHash] of TOKEN_PAIRS) {
    let [_err, results] = await asyncWrap(getTokenInfo(net, scriptHash)) // eslint-disable-line
    if (results) {
      tokensInfo.push({
        symbol,
        info: results
      })
    }
  }
  return dispatch(setTokensInfo(tokensInfo))
}

// state getters
export const getNeo = (state) => state.wallet.NEO
export const getGas = (state) => state.wallet.GAS
export const getTransactions = (state) => state.wallet.transactions
export const getNeoPrice = (state) => state.wallet.neoPrice
export const getGasPrice = (state) => state.wallet.gasPrice
export const getTokens = (state) => state.wallet.tokens

const initialState = {
  NEO: 0,
  GAS: 0,
  transactions: [],
  neoPrice: 0,
  gasPrice: 0,
  tokens: []
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case SET_BALANCE:
      const { Neo, Gas } = action.payload
      return {
        ...state,
        NEO: Neo,
        GAS: Gas
      }
    case SET_NEO_PRICE:
      const { neoPrice } = action.payload
      return {
        ...state,
        neoPrice
      }
    case SET_GAS_PRICE:
      const { gasPrice } = action.payload
      return {
        ...state,
        gasPrice
      }
    case RESET_PRICE:
      return {
        ...state,
        neoPrice: 0,
        gasPrice: 0
      }
    case SET_TRANSACTION_HISTORY:
      const { transactions } = action.payload
      return {
        ...state,
        transactions
      }
    case SET_TOKENS_BALANCE:
      const { tokens } = action.payload
      return {
        ...state,
        tokens: merge(state.tokens, tokens)
      }
    case SET_TOKENS_INFO:
      const { tokensInfo } = action.payload
      return {
        ...state,
        tokens: merge(state.tokens, tokensInfo)
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
