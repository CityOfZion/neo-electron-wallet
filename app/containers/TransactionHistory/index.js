// @flow
import { connect, type MapStateToProps } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'

import withNetworkData from '../../hocs/withNetworkData'
import withAuthData from '../../hocs/withAuthData'
import { syncTransactionHistory, getIsLoadingTransactions } from '../../modules/transactions'
import { getTransactions } from '../../modules/wallet'
import TransactionHistory from './TransactionHistory'

const mapStateToProps: MapStateToProps<*, *, *> = (state: Object) => ({
  transactions: getTransactions(state),
  isLoadingTransactions: getIsLoadingTransactions(state)
})

const actionCreators = {
  syncTransactionHistory
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withNetworkData(),
  withAuthData()
)(TransactionHistory)
