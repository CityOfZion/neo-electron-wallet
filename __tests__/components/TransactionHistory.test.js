import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import { setTransactionHistory } from '../../app/modules/wallet'

import TransactionHistory from '../../app/containers/TransactionHistory'

const initialState = {
  account: {
    loggedIn: true,
    wif: 'L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou',
    address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    transactions: []
  }
}

const transactions = {
  wallet: {
    transactions: [
      {
        type: 'NEO',
        amount: '50',
        txid: '76938979'
      },
      {
        type: 'GAS',
        amount: '0.4',
        txid: '76938980'
      }
    ]
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<TransactionHistory store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <TransactionHistory />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('TransactionHistory', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  test('calls syncTransactionHistory after rendering', (done) => {
    const { store } = setup(initialState, false)
    const state = store.getState()
    setTimeout(() => {
      expect(store.getActions()[0]).toEqual(setTransactionHistory(initialState.wallet.transactions))
      done()
    }, 0)
  })

  test('correctly renders no transaction history', (done) => {
    const { store, wrapper } = setup(initialState, false)

    const columnHeader = wrapper.find('.columnHeader')
    expect(columnHeader.text()).toEqual('Transaction History')

    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(0)
    done()
  })

  test('correctly renders with NEO and GAS transaction history', (done) => {
    const transactionState = Object.assign({}, initialState, transactions)
    const { store, wrapper } = setup(transactionState, false)

    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(2)
    expect(transactionList.childAt(0).find('.txid').text()).toEqual(transactions.wallet.transactions[0].txid)
    expect(transactionList.childAt(1).find('.txid').text()).toEqual(transactions.wallet.transactions[1].txid)
    expect(transactionList.childAt(0).find('.amount').text()).toEqual(`${transactions.wallet.transactions[0].amount} ${transactions.wallet.transactions[0].type}`)
    expect(transactionList.childAt(1).find('.amount').text()).toEqual(`${transactions.wallet.transactions[1].amount} ${transactions.wallet.transactions[1].type}`)
    done()
  })
})
