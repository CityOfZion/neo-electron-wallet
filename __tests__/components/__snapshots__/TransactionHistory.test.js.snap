// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHistory renders without crashing 1`] = `
<Connect(withData(mapProps(Connect(withData(TransactionHistory)))))
  isLoadingTransactions={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  syncTransactionHistory={[Function]}
  transactions={Array []}
/>
`;
